### PR TITLE
Experiment descriptor

### LINKED ISSUE #3: Experiment descriptor
Just like [msgpush-go](https://github.com/gcinterceptor/msgpush-go) or [msgpush-ruby](https://github.com/gcinterceptor/msgpush-ruby), this version of msgpush needs a descriptor for experiments.

### PR DESCRIPTION
**Pull request by dfquaresma**
This PR fix #3.

### CODE REVIEW

File: experiments/config.toml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+nginx_version = "nginx/1.14.0 (Ubuntu)"
+so_version = "Ubuntu 18.04 LTS"
+kernel_version = "4.15.0-29-generic"
+
+[execution]
+# exports: export commands which should be used to export common environment variables. Should use outdir=stateless for stateless experiments and outdir=stateful for stateful experiments.
+exports = 'export outdir="stateless"; export expdir=${PWD}'
+
+# killall: kill all processes started by the experiment.
+killall = 'for ip in $ip1 $ip2 $ip3 $ip4; do ssh -i ${sshkey} ubuntu@${ip} "killall java 2>/dev/null; rm gc.log shed_${expid}_${ip}.csv 2>/dev/null; killall mon.sh 2>/dev/null; killall gci-proxy 2>/dev/null; rm proxy.* 2>/dev/null;"; done'
```
Comment (danielfireman) [inline experiments/config.toml]: killall should not do file cleanups if it is strictly necessary.

> rm gc.log shed_${expid}_${ip}.csv 2>/dev/null;
> rm proxy.* 2>/dev/null;

File: experiments/config.toml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+# load: trigger the load.
+load = 'ssh -i ${sshkey} ubuntu@${lb} "sudo rm /var/log/nginx/*.log; sudo service nginx restart; killall vegeta >/dev/null 2>&1; echo \"GET http://127.0.0.1/\" | ./vegeta attack -duration=120s -rate=30 | tee results.bin | ./vegeta report >client_${expid}.out 2>client_${expid}.err";'
+
+# fetch_results: fetches experiment results.
+fetch_results = 'scp -i ${sshkey} ubuntu@${lb}:~/*_${expid}* ${outdir}; for ip in $ip1 $ip2 $ip3 $ip4; do scp -i ${sshkey} ubuntu@${ip}:~/*_${expid}_${ip}* ${outdir}; done'
+
+# cleanup: killall processes, removes files and revert configuration changes done by the experiment run.
+cleanup = 'ssh -i ${sshkey} ubuntu@${lb} "rm *${expid}*"; for ip in $ip1 $ip2 $ip3 $ip4; do ssh -i ${sshkey} ubuntu@${ip} "killall java; killall mon.sh; killall gci-proxy 2>/dev/null; rm proxy.* 2>/dev/null; rm *${expid}_${ip}*"; done'
```
Comment (danielfireman) [inline experiments/config.toml]: cleanup should not kill processes (killall does)

> killall java; killall mon.sh; killall gci-proxy 2>/dev/null

File: experiments/config.toml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+setup = 'git clone http://github.com/gcinterceptor/gci-java && cd gci-java/ && ./build.sh; mv core/src/main/java/libgc.so ${expdir}; cd ${expdir}; git clone http://github.com/gcinterceptor/msgpush-java && cd msgpush-java && mvn clean package; cp experiments/mon.sh ${expdir}; cd target/ && mv msgpush-0.0.1-SNAPSHOT.jar msgpush-java.jar; cp msgpush-java.jar ${expdir}; cd ${expdir}; for ip in $ip1 $ip2 $ip3 $ip4; do scp -i ${sshkey} mon.sh ubuntu@${ip}:~/; scp -i ${sshkey} msgpush-java.jar ubuntu@${ip}:~/; scp -i ${sshkey} libgc.so ubuntu@${ip}:~/; done; go get -u github.com/gcinterceptor/gci-proxy; go get -u github.com/julienschmidt/httprouter; cd $GOPATH/src/github.com/gcinterceptor/gci-proxy && go build; for ip in $ip1 $ip2 $ip3 $ip4; do scp -i ${sshkey} gci-proxy ubuntu@${ip}:~/; done; cd ${expdir}'
+
+# load: trigger the load.
+load = 'ssh -i ${sshkey} ubuntu@${lb} "sudo rm /var/log/nginx/*.log; sudo service nginx restart; killall vegeta >/dev/null 2>&1; echo \"GET http://127.0.0.1/\" | ./vegeta attack -duration=120s -rate=30 | tee results.bin | ./vegeta report >client_${expid}.out 2>client_${expid}.err";'
```
Comment (danielfireman) [inline experiments/config.toml]: Here we should have the warmup. Please refer to msgpush-go experiment:

> \#load: trigger the load.
load = 'ssh -i ${sshkey} ubuntu@${lb} "sudo rm /var/log/nginx/access.log; sudo /etc/init.d/nginx restart; echo \"GET http://localhost/\" | ./vegeta attack -duration=60s -rate=80 >/dev/null; sudo rm /var/log/nginx/access.log; sudo /etc/init.d/nginx restart"; for ip in $ip1 $ip2 $ip3 $ip4; do ssh -i ${sshkey} ubuntu@${ip} "truncate -s 0 *${expid}_${ip}*"; done; ssh -i ${sshkey} ubuntu@${lb} "echo \"GET http://localhost/\" | ./vegeta attack -duration=120s -rate=80 > results.bin; cat results.bin | ./vegeta report > vegeta_${expid}.out; cat results.bin | ./vegeta report -reporter=plot > plot_${expid}.html; rm results.bin; cp /var/log/nginx/access.log al_${expid}.csv"'

File: experiments/config.toml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        exports = 'export expid="nogci_java_3"'
+        
+        [execution.no_gci.4]
+        exports = 'export expid="nogci_java_4"'
+        
+        [execution.no_gci.5]
+        exports = 'export expid="nogci_java_5"'
+
+    [execution.gci]
+    start = 'for ip in $ip1 $ip2 $ip3 $ip4; do ssh -i ${sshkey} ubuntu@${ip} "killall java 2>/dev/null; rm gc.log shed.csv 2>/dev/null; killall mon.sh 2>/dev/null; USE_GCI=true WINDOW_SIZE=1 MSG_SIZE=262144 YOUNG_GEN=256000000 nohup java -Djvmtilib=~/libgc.so -server -Xms512m -Xmx512m -Xlog:gc:file=gc.log -XX:+UseG1GC -XX:+ExplicitGCInvokesConcurrent -XX:+UnlockExperimentalVMOptions -XX:G1NewSizePercent=50 -XX:G1MaxNewSizePercent=50 -Dserver.port=8080 -jar  msgpush-java.jar  >msgpush-java_${expid}_${ip}.out 2>msgpush-java_${expid}_${ip}.err & nohup ./mon.sh >cpu_${expid}_${ip}.csv 2>/dev/null &"; done; ssh -i ${sshkey} ubuntu@${ip} "killall gci-proxy 2>/dev/null; rm proxy.* 2>/dev/null; ./gci-proxy --port 3000 --url http://localhost:8080 --ygen=67108864 --tgen=6710886 >proxy_${expid}_${ip}.out 2>proxy_${expid}_${ip}.err &";'
```
Comment (danielfireman) [inline experiments/config.toml]: No need to kill or do any cleanups here (yes, we need to fix msgpush-go as well).

File: experiments/config.toml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+fetch_results = 'scp -i ${sshkey} ubuntu@${lb}:~/*_${expid}* ${outdir}; for ip in $ip1 $ip2 $ip3 $ip4; do scp -i ${sshkey} ubuntu@${ip}:~/*_${expid}_${ip}* ${outdir}; done'
+
+# cleanup: killall processes, removes files and revert configuration changes done by the experiment run.
+cleanup = 'ssh -i ${sshkey} ubuntu@${lb} "rm *${expid}*"; for ip in $ip1 $ip2 $ip3 $ip4; do ssh -i ${sshkey} ubuntu@${ip} "killall java; killall mon.sh; killall gci-proxy 2>/dev/null; rm proxy.* 2>/dev/null; rm *${expid}_${ip}*"; done'
+
+    [execution.no_gci]
+    # start: start all processes needed by the experiment run. Update some values to execute correctly the experiment.
+    start = 'for ip in $ip1 $ip2 $ip3 $ip4; do ssh -i ${sshkey} ubuntu@${ip} "killall java 2>/dev/null; rm gc.log shed.csv 2>/dev/null; killall mon.sh 2>/dev/null; USE_GCI=false WINDOW_SIZE=1 MSG_SIZE=262144 YOUNG_GEN=256000000 nohup java -server -Xms512m -Xmx512m -Xlog:gc:file=gc.log  -XX:+UseG1GC -XX:+DisableExplicitGC -Dserver.port=3000 -jar  msgpush-java.jar  >msgpush-java_${expid}_${ip}.out 2>msgpush-java_${expid}_${ip}.err & nohup ./mon.sh >cpu_${expid}_${ip}.csv 2>/dev/null &"; done;'
```
Comment (danielfireman) [inline experiments/config.toml]: No need to kill or do any cleanups here.

File: experiments/config.toml (hunk 4 above)
Comment (danielfireman) [inline experiments/config.toml]: `WINDOW_SIZE=0` ?

**Review by danielfireman — CHANGES_REQUESTED** [review]

File: experiments/config.toml (hunk 4 above)
Comment (dfquaresma) [reply experiments/config.toml]: Sorry. My bad.

**Comment by dfquaresma** [comment]
Thanks for review @danielfireman! I'm working on the requested changes

File: experiments/config.toml (hunk 1 above)
Comment (dfquaresma) [reply experiments/config.toml]: Done

File: experiments/config.toml (hunk 2 above)
Comment (dfquaresma) [reply experiments/config.toml]: Done

File: experiments/config.toml (hunk 3 above)
Comment (dfquaresma) [reply experiments/config.toml]: Done

File: experiments/config.toml (hunk 4 above)
Comment (dfquaresma) [reply experiments/config.toml]: Done

File: experiments/config.toml (hunk 5 above)
Comment (dfquaresma) [reply experiments/config.toml]: Done

File: experiments/config.toml (hunk 4 above)
Comment (dfquaresma) [reply experiments/config.toml]: Done

File: experiments/config.toml (hunk 3 above)
Comment (danielfireman) [reply experiments/config.toml]: Not done. Please do.

**Review by danielfireman — CHANGES_REQUESTED** [review]

**Review by danielfireman — APPROVED** [review]

**Pull request merged**